src/Cmfd.cpp

```diff
@@ -832,48 +832,50 @@ int Cmfd::getCmfdGroup(int group) {
 
 
 /**
- * @brief Set the CMFD energy group structure.
+ * @brief Set a coarse energy group structure for CMFD.
  * @details CMFD does not necessarily need to have the same energy group
  *          structure as the MOC problem. This function can be used to set
- *          a sparse energy group structure to speed up the CMFD solve.
- * @param group_indices An array of the CMFD group boundaries
- * @param length_group_indices The length of the group_indices array
+ *          a sparse energy group structure to speed up the CMFD solve. An
+ *          example of how this may be called from Python to use a coarse
+ *          2-group CMFD structure atop a fine 7-group MOC structure is
+ *          illustrated below:
+ *
+ * @code
+ *          cmfd.setGroupStructure([[1,2,3], [4,5,6,7]])
+ * @endcode
+ *
+ * @param group_indices A nested vector of MOC-to-CMFD group mapping
  */
-void Cmfd::setGroupStructure(int* group_indices, int length_group_indices) {
+void Cmfd::setGroupStructure(std::vector< std::vector<int> > group_indices) {
 
-  _num_cmfd_groups = length_group_indices - 1;
+  _user_group_indices = true;
 
   /* Delete old group indices array if it exists */
   if (_group_indices != NULL)
     delete [] _group_indices;
 
   /* Allocate memory for new group indices */
-  _group_indices = new int[length_group_indices];
-
-  if (group_indices == NULL) {
-    for (int i = 0; i < length_group_indices; i++) {
-      _group_indices[i] = i;
-    }
-  }
-  else{
-    if (group_indices[0] != 1)
-      log_printf(ERROR, "The first value in group indices must be 1!");
+  _num_cmfd_groups = group_indices.size();
+  _group_indices = new int[_num_cmfd_groups+1];
 
-    /* Set first group indice to 0 */
-    _group_indices[0] = 0;
+  if (group_indices[0][0] != 1)
+    log_printf(WARNING, "The first CMFD coarse group index is %d but is being "
+	       "over-ridden to 1", group_indices[0][0], group_indices[0][0]);
 
-    /* Set MOC group bounds for rest of CMFD energy groups */
-    for (int i = 1; i < length_group_indices; i++) {
-      /* Check that the group indices are always increasing */
-      if (group_indices[i] <= group_indices[i-1])
-        log_printf(ERROR, "The group indices must be increasing!");
+  /* Set first group index to 0 */
+  _group_indices[0] = 0;
 
-      _group_indices[i] = group_indices[i] - 1;
-      log_printf(DEBUG, "group indices %d: %d", i, group_indices[i]);
-    }
+  /* Set MOC group bounds for rest of CMFD energy groups */
+  for (int i=1; i < _num_cmfd_groups; i++) {
+    _group_indices[i] = group_indices[i][0] - 1;
+    log_printf(DEBUG, "group indices %d: %d", i, _group_indices[i]);
```

Comment:
Could we have a check to make sure the group indices are monotonically increasing? Most users will probably know very little about CMFD and this is one of the few user inputs, so I think it would be nice to have an error check.